File: tests/test_scheduler.py
"""Scheduler tests."""
from scheduler import Scheduler


def test_empty_class():
    """Tests all conditions being empty."""
    scheduler = Scheduler()
    assert scheduler.create_schedule() == []


def test_1xMax59min():
    """Tests 1 event of at most 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.fit_interval = 1
    scheduler.items = {"chem"}
    scheduler.event_count = {("chem", 1, 30, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 30)]
    scheduler.event_count = {("chem", 1, 45, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 45)]
    scheduler.event_count = {("chem", 1, 59, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 59)]


def test_1xMax1439min():
    """Tests 1 event of at most 23 hours and 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.fit_interval = 1
    scheduler.items = {"chem"}
    scheduler.event_count = {("chem", 1, 60, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 60)]
    scheduler.event_count = {("chem", 1, 87, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 87)]
    scheduler.event_count = {("chem", 1, 119, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 119)]
    scheduler.event_count = {("chem", 1, 453, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 453)]
    scheduler.event_count = {("chem", 1, 1439, "per month")}
    assert scheduler.create_schedule() == [("chem", 0, 1439)]


def test_1xPerTuesday():
    """Tests 1 event with 1 occurence per weekday."""
    scheduler = Scheduler()
    scheduler.fit_interval = 10
    scheduler.items = {"chem"}
    scheduler.event_count = {("chem", 1, 30, "per tuesday")}
    assert scheduler.create_schedule() == [
        ('chem', 2880, 2910), ('chem', 12960, 12990),
        ('chem', 23040, 23070), ('chem', 33120, 33150)]


def test_1xPerWeek():
    """Tests 1 event with 1 occurence per weekday."""
    scheduler = Scheduler()
    scheduler.fit_interval = 10
    scheduler.items = {"chem"}
    scheduler.event_count = {("chem", 1, 30, "per week")}
    assert scheduler.create_schedule() == [
        ('chem', 0, 30), ('chem', 10080, 10110),
        ('chem', 20160, 20190), ('chem', 30240, 30270)]


def test_1xPerMonth():
    """Tests 1 event with 1 occurence per weekday."""
    scheduler = Scheduler()
    scheduler.fit_interval = 100
    scheduler.items = {"pt"}
    scheduler.event_count = {("pt", 2, 300, "per month")}
    assert scheduler.create_schedule() == [('pt', 0, 300), ('pt', 300, 600)]


def test_1xPerDay():
    """Tests 1 event with 1 occurence per weekday."""
    scheduler = Scheduler()
    scheduler.items = {"chem"}
    scheduler.event_count = {("chem", 1, 30, "per day")}
    scheduler.fit_interval = 10
    assert scheduler.create_schedule() == [
        ("chem", 1440 * i, (1440 * i) + 30) for i in range(28)]


def test_2xMax59min():
    """Tests 2 event of at most 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.fit_interval = 1
    scheduler.items = {"chem", "math"}
    scheduler.event_count = [
        ("chem", 1, 30, "per month"), ("math", 1, 45, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 30), ("math", 30, 75)]
    scheduler.event_count = [
        ("chem", 1, 1, "per month"), ("math", 1, 45, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 1), ("math", 1, 46)]
    scheduler.event_count = [
        ("chem", 1, 59, "per month"), ("math", 1, 45, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 59), ("math", 59, 104)]
    scheduler.event_count = [
        ("chem", 1, 59, "per month"), ("math", 1, 59, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 59), ("math", 59, 118)]


def test_2xMax1439min():
    """Tests 2 event of at most 23 hours and 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.fit_interval = 5
    scheduler.items = {"chem", "math"}
    scheduler.event_count = [
        ("chem", 1, 60, "per month"), ("math", 1, 45, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 60), ("math", 60, 105)]
    scheduler.event_count = [
        ("chem", 1, 160, "per month"), ("math", 1, 35, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 160), ("math", 160, 195)]
    scheduler.event_count = [
        ("chem", 1, 235, "per month"), ("math", 1, 165, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 235), ("math", 235, 400)]


def test_3xMax59min():
    """Tests 3 event of at most 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.fit_interval = 15
    scheduler.items = {"chem", "math"}
    scheduler.event_count = [
        ("chem", 2, 30, "per month"), ("math", 1, 45, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 30),
        ("chem", 30, 60),
        ("math", 60, 105)]


def test_3xMax1439min():
    """Tests 3 event of at most 23 hours and 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.items = {"chem", "math"}
    scheduler.fit_interval = 50
    scheduler.event_count = [
        ("chem", 2, 300, "per month"), ("math", 1, 200, "per month")]
    assert scheduler.create_schedule() == [
        ("chem", 0, 300),
        ("chem", 300, 600),
        ("math", 600, 800)]


def test_cleanBlockedTimeLogic():
    """Tests the function cleanBlockedTimeLogic."""
    scheduler = Scheduler()
    blocks = []
    assert scheduler.cleanBlockedTimeLogic(blocks) == []
    blocks = [
        ('BLOCKED', 17, 35), ('BLOCKED', 20, 45),
        ('BLOCKED', 0, 10), ('BLOCKED', 60, 100)]
    assert scheduler.cleanBlockedTimeLogic(blocks) == [
        ('BLOCKED', 0, 10), ('BLOCKED', 17, 45), ('BLOCKED', 60, 100)]
    blocks = [
        ('BLOCKED', 5, 35), ('BLOCKED', 20, 45),
        ('BLOCKED', 2, 10), ('BLOCKED', 40, 100)]
    assert scheduler.cleanBlockedTimeLogic(blocks) == [
        ('BLOCKED', 2, 100)]


def test_blockedTime():
    """Tests period blocking functionality."""
    # One block everyday
    scheduler = Scheduler()
    scheduler.items = {"chem", "math"}
    scheduler.fit_interval = 5
    scheduler.event_count = [("math", 1, 55, "per month")]
    scheduler.blocked_times = {(0, 0, 3, 0, "every day")}
    ans = [('BLOCKED', a*1440, (a*1440)+(60*3)) for a in range(28)]
    ans.insert(1, ('math', 180, 235))
    assert scheduler.create_schedule() == ans
    # One block every sunday
    scheduler.event_count = [("math", 1, 55, "per month")]
    scheduler.blocked_times = {(0, 0, 23, 20, "every sunday")}
    ans = [('BLOCKED', 10080*a, (10080*a) + (23*60) + 20) for a in range(4)]
    ans.insert(1, ('math', 1400, 1455))
    assert scheduler.create_schedule() == ans


def test_afterLogic():
    """Tests event logic functionality."""
    scheduler = Scheduler()
    scheduler.items = {"chem", "math"}
    scheduler.logic = {("break", "after", "math")}
    scheduler.fit_interval = 1
    scheduler.event_count = [
        ("break", 1, 40310, "per month"),
        ("math", 1, 5, "per month")]
    assert scheduler.create_schedule() == [
        ('math', 0, 5), ('break', 5, 40315)]
    scheduler.event_count = [
        ("break", 1, 15, "per month"),
        ("math", 1, 5, "per month")]
    assert scheduler.create_schedule() == [
        ('break', 0, 15), ('math', 1440, 1445)]
    scheduler.logic = {("math", "after", "break")}
    scheduler.event_count = [
        ("break", 1, 40, "per month"),
        ("math", 1, 5, "per month")]
    assert scheduler.create_schedule() == [
        ('break', 0, 40), ('math', 40, 45)]


def test_beforeLogic():
    """Tests event logic functionality."""
    scheduler = Scheduler()
    scheduler.items = {"chem", "math"}
    scheduler.logic = {("break", "before", "math")}
    scheduler.fit_interval = 1
    scheduler.event_count = [
        ("break", 1, 5, "per month"),
        ("math", 1, 40, "per month")]
    assert scheduler.create_schedule() == [
        ('break', 0, 5), ('math', 5, 45)]
    scheduler.logic = {("math", "before", "break")}
    scheduler.event_count = [
        ("break", 1, 40200, "per month"),
        ("math", 1, 40, "per month")]
    assert scheduler.create_schedule() == [
        ('math', 0, 40), ('break', 40, 40240)]


def test_firstEvent():
    # scheduler = Scheduler()
    # scheduler.items = {"chem", "math"}
    # scheduler.event_count = {
    #     ("chem", 1, 30, "per month"), ("math", 1, 30, "per month")}    
    # scheduler.first_events["sunday"] = ["chem"]
    # scheduler.fit_interval = 15
    # scheduler.optimize()
    # schedule = [
    #     ("START", -1, 0), ('BLOCKED', 90, 40315), ("END", 40320, 40321)]
    # scheduler.backtrack(schedule, [('math', 30), ('chem', 30)], 0)
    # assert scheduler.schedule[1:-1] == [
    #     ('chem', 0, 30), ('math', 30, 60), ('BLOCKED', 90, 40315)]
    # scheduler.first_events["sunday"] = ["math"]
    # scheduler.backtrack(schedule, [('math', 30), ('chem', 30)], 0)
    # assert scheduler.schedule[1:-1] == [
    #     ('math', 0, 30), ('chem', 30, 60), ('BLOCKED', 90, 40315)]


    scheduler = Scheduler()
    scheduler.items = {"chem", "math"}
    scheduler.event_count = [
        ("chem", 2, 100, "per month"), ("math", 1, 100, "per month")]   
    scheduler.first_events["sunday"] = ["math"]
    scheduler.fit_interval = 100
    assert scheduler.create_schedule() == [
        ('math', 0, 100), ('chem', 100, 200), ('chem', 200, 300)]



 

def test_lastEvent():
    scheduler = Scheduler()
    assert False

def test_eventTimes():
    scheduler = Scheduler()
    assert False

File: test/test_scheduler.jl

using TestItems

@testitem "Empty" begin
    """Tests most basic case (no data passed)."""
    scheduler = Scheduler()
    @test create_schedule(scheduler) == []
end

@testitem "1xMax59min" begin
    """Tests 1 event of at most 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 30, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 30)]
    scheduler.eventCount = Set([("chem", 1, 45, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 45)]
    scheduler.eventCount = Set([("chem", 1, 59, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 59)]
end

@testitem "1xMax1439min" begin
    """Tests 2 event of at most 23 hours and 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 60, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 60)]
    scheduler.eventCount = Set([("chem", 1, 87, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 87)]
    scheduler.eventCount = Set([("chem", 1, 119, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 119)]
    scheduler.eventCount = Set([("chem", 1, 453, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 453)]
    scheduler.eventCount = Set([("chem", 1, 1439, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 1439)]
end

@testitem "1xPerTuesday" begin
    """Tests 1 event with 1 occurence per weekday."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 30, "per tuesday")])
    @test create_schedule(scheduler) == [
        ("chem", 2880, 2910), ("chem", 12960, 12990),
        ("chem", 23040, 23070), ("chem", 33120, 33150)]
end

@testitem "1xPerWeek" begin
    """Tests 1 event with 1 occurence per week."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 30, "per week")])
    @test create_schedule(scheduler) == [
        ("chem", 0, 30), ("chem", 10080, 10110),
        ("chem", 20160, 20190), ("chem", 30240, 30270)]
end

@testitem "1xPerMonth" begin
    """Tests 1 event with 1 occurence per month."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("pt", 2, 300, "per month")])
    @test create_schedule(scheduler) == [("pt", 0, 300), ("pt", 300, 600)]
end

@testitem "1xPerDay" begin
    """Tests 1 event with 1 occurence per weekday."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 30, "per day")])
    @test create_schedule(scheduler) == [
        ("chem", 1440 * i, (1440 * i) + 30) for i in range(0, 27)]
end

@testitem "2xMax59min" begin
    """Tests 2 event of at most 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 30, "per month"), ("math", 1, 45, "per month")])
    @test create_schedule(scheduler) ∈ [[("chem", 0, 30), ("math", 30, 75)],[("math", 0, 45), ("chem", 45, 75)]]
    scheduler.eventCount = Set([("chem", 1, 1, "per month"), ("math", 1, 45, "per month")])
    @test create_schedule(scheduler) ∈ [[("chem", 0, 1), ("math", 1, 46)],[("math", 0, 45), ("chem", 45, 46)]]
    scheduler.eventCount = Set([("chem", 1, 59, "per month"), ("math", 1, 45, "per month")])
    @test create_schedule(scheduler) ∈ [[("math", 0, 45), ("chem", 45, 104)],[("chem", 0, 59), ("math", 59, 104)]]
    scheduler.eventCount = Set([("chem", 1, 59, "per month"), ("math", 1, 59, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 59), ("math", 59, 118)]
end

@testitem "2xMax1439min" begin
    """Tests 2 event of at most 23 hours and 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 1, 60, "per month"), ("math", 1, 45, "per month")])
    @test create_schedule(scheduler) ∈ [[("math", 0, 45), ("chem", 45, 105)],[("chem", 0, 60), ("math", 60, 105)]]
    scheduler.eventCount = Set([("chem", 1, 160, "per month"), ("math", 1, 35, "per month")])
    @test create_schedule(scheduler) ∈ [[("math", 0, 35), ("chem", 35, 195)],[("chem", 0, 160), ("math", 160, 195)]]
    scheduler.eventCount = Set([("chem", 1, 235, "per month"), ("math", 1, 165, "per month")])
    @test create_schedule(scheduler) ∈ [[("math", 0, 165), ("chem", 165, 400)],[("chem", 0, 235), ("math", 235, 400)]]
end

@testitem "3xMax59min" begin
    """Tests 3 event of at most 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 2, 30, "per month"), ("math", 1, 45, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 30), ("chem", 30, 60), ("math", 60, 105)]
end

@testitem "3xMax1439min" begin
    """Tests 3 event of at most 23 hours and 59 minutes duration."""
    scheduler = Scheduler()
    scheduler.eventCount = Set([("chem", 2, 300, "per month"), ("math", 1, 200, "per month")])
    @test create_schedule(scheduler) == [("chem", 0, 300), ("chem", 300, 600), ("math", 600, 800)]
end

@testitem "cleanEvents" begin
    """Tests the cleanBlockedTimeLogic() function."""
    scheduler = Scheduler()
    @test cleanEvents(Vector{Tuple{String,Int64,Int64}}()) == []
    blocks = [
        ("BLOCKED", 0, 10), ("BLOCKED", 17, 35), 
        ("BLOCKED", 20, 45), ("BLOCKED", 60, 100)]
    @test cleanEvents(blocks) == [
        ("BLOCKED", 0, 10), ("BLOCKED", 17, 45), ("BLOCKED", 60, 100)]
    blocks = [
        ("BLOCKED", 2, 10), ("BLOCKED", 5, 35), 
        ("BLOCKED", 20, 45), ("BLOCKED", 40, 100)]
    @test cleanEvents(blocks) == [
        ("BLOCKED", 2, 100)]
end

@testitem "blockedTime" begin
    """Tests blocked times functionality."""
    scheduler = Scheduler()
    # One block everyday
    scheduler.eventCount = Set([("math", 1, 55, "per month")])
    scheduler.blockedTimes = Set(Set([(0, 0, 3, 0, "every day")]))
    ans = [("BLOCKED", a*1440, (a*1440)+(60*3)) for a in range(0, 27)]
    insert!(ans, 2, ("math", 180, 235))
    @test create_schedule(scheduler) == ans
    # One block every sunday
    scheduler.eventCount = Set([("math", 1, 55, "per month")])
    scheduler.blockedTimes = Set([(0, 0, 23, 20, "every sunday")])
    ans = [("BLOCKED", 10080*a, (10080*a) + (23*60) + 20) for a in range(0, 3)]
    insert!(ans, 2, ("math", 1400, 1455))
    @test create_schedule(scheduler) == ans
end

@testitem "1xAfterLogic" begin
    """Tests event logic functionality with 1 after-based logic."""
    scheduler = Scheduler()
    scheduler.logic = Set([("break", "after", "math")])
    scheduler.eventCount = Set([("break", 1, 15, "per month"), ("math", 1, 5, "per month")])
    @test create_schedule(scheduler) == [("math", 0, 5), ("break", 5, 20)]
    scheduler.eventCount = Set([("break", 1, 40310, "per month"), ("math", 1, 5, "per month")])
    @test create_schedule(scheduler) == [("math", 0, 5), ("break", 5, 40315)]
    scheduler.logic = Set([("math", "after", "break")])
    scheduler.eventCount = Set([("break", 1, 40, "per month"), ("math", 1, 5, "per month")])
    @test create_schedule(scheduler) ∈ [[("break", 0, 40), ("math", 40, 45)]]
end

@testitem "1xBeforeLogic" begin
    """Tests event logic functionality with 1 before-based logic."""
    scheduler = Scheduler()
    scheduler.logic = Set([("break", "before", "math")])
    scheduler.eventCount = Set([("break", 1, 5, "per month"), ("math", 1, 40, "per month")])
    @test create_schedule(scheduler) ∈ [[("break", 0, 5), ("math", 5, 45)],[("math", 0, 40), ("break", 1440, 1445)]]
    scheduler.logic = Set([("math", "before", "break")])
    scheduler.eventCount = Set([("break", 1, 40200, "per month"), ("math", 1, 40, "per month")])
    @test create_schedule(scheduler) == [("math", 0, 40), ("break", 40, 40240)]
end

@testitem "1x1x1EventTimes" begin
    """Tests the EventTimes functionality with 1 event, 1 rule and 1 day."""
    scheduler = Scheduler()
    scheduler.eventTimes = Set([("math", 2, 30, 4, 0, "every day")])
    scheduler.eventCount = Set([("math", 1, 30, "per month")])
    @test create_schedule(scheduler) == [("math", 150, 180)]
    scheduler.eventCount = Set([("math", 1, 91, "per month")])
    @test create_schedule(scheduler) == []
    scheduler.eventCount = Set([("math", 1, 30, "per month")])
    scheduler.eventTimes = Set([("math", 2, 30, 4, 0, "every sunday")])
    @test create_schedule(scheduler) == [("math", 150, 180)]
    scheduler.eventTimes = Set([("math", 2, 30, 4, 0, "every monday")])
    @test create_schedule(scheduler) == [("math", 1590, 1620)]
    scheduler.eventTimes = Set([("math", 14, 30, 15, 0, "every day")])
    @test create_schedule(scheduler) == [("math", 870, 900)]
end

@testitem "1x2x1EventTimes" begin
    """Tests the EventTimes functionality with 1 event, 2 rules and 1 day."""
    scheduler = Scheduler()
    scheduler.eventTimes = Set([("math", 2, 30, 2, 45, "every day"),("math", 5, 0, 5, 30, "every day")])
    scheduler.eventCount = Set([("math", 1, 30, "per month")])
    @test create_schedule(scheduler) == [("math", 300, 330)]
end

@testitem "2x1x1EventTimes" begin
    """Tests the EventTimes functionality with 2 events, 1 rule and 1 day."""
    scheduler = Scheduler()
    scheduler.eventTimes = Set([("math", 0, 0, 1, 30, "every day")])
    scheduler.eventCount = Set([("math", 2, 30, "per month")])
    @test create_schedule(scheduler) == [("math", 0, 30),("math", 30, 60)]
end

@testitem "2x1x2CheckEventTimes" begin
    """Tests the checkEventTimes() function with 2 events, 1 rule and 2 days."""
    scheduler = Scheduler()
    scheduler.eventTimes = Set([("math", 23, 30, 1, 30, "every day")])
    scheduler.eventCount = Set([("math", 29, 60, "per month")])
    @test checkEventTimes([("START", -1, 0),("math", 1410, 1470), 
        ("math", 1470, 1530),("END", 40320, 40321)],2,scheduler) == false
end

@testitem "fillEmptyGaps" begin
    """Tests the fillEmptyGaps() function."""
    scheduler = Scheduler()
    scheduler.schedule = []
    @test fillEmptyGaps(scheduler.schedule) == [("", 0, 40320)]
    scheduler.schedule = [("math", 0, 1250), ("chem", 1305, 1440)]
    @test fillEmptyGaps(scheduler.schedule) == [
        ("math", 0, 1250),("", 1250, 1305),("chem", 1305, 1440),("", 1440, 40320)]
end

@testitem "separateMultidayEvents" begin
    """Tests the separateMultidayEvents() function."""
    scheduler = Scheduler()
    @test separateMultidayEvents(scheduler.schedule) == []
    @test separateMultidayEvents([("math", 0, 1000)]) == [("math", 0, 1000)]
    @test separateMultidayEvents([("math", 0, 1440)]) == [("math", 0, 1440)]
    @test separateMultidayEvents([("math", 0, 1500)]) == [
        ("math", 0, 1440),("math", 1440, 1500)]
    @test separateMultidayEvents([("math", 0, 3000)]) == [
        ("math", 0, 1440),("math", 1440, 2880),("math", 2880, 3000)]
    @test separateMultidayEvents([("math", 0, 1500),("chem", 1500, 3000)]) == [
        ("math", 0, 1440),("math", 1440, 1500),("chem", 1500, 2880),("chem", 2880, 3000)]
end
